include/zephyr/drivers/charger.h

```diff
@@ -0,0 +1,115 @@
+/*
+ * Copyright 2023 Cirrus Logic, Inc.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_INCLUDE_DRIVERS_CHARGER_H_
+#define ZEPHYR_INCLUDE_DRIVERS_CHARGER_H_
+
+#ifdef __cplusplus
+extern "C" {
+#endif /* __cplusplus */
+
+#include <stdbool.h>
+#include <stddef.h>
+#include <stdint.h>
+
+#include <zephyr/device.h>
+
+/* Keep these alphabetized wrt property name */
+
+/** Runtime Dynamic Battery Parameters */
+
+/** Represents the health of the charger */
+#define CHARGER_HEALTH          0
+/** Indicates if VBUS is present for the charger */
+#define CHARGER_ONLINE          CHARGER_HEALTH + 1
+/** Reports whether or not a battery is present */
+#define CHARGER_PRESENT         CHARGER_ONLINE + 1
+/** Represents the charging status of the charger */
+#define CHARGER_STATUS          CHARGER_PRESENT + 1
+
+/** Reserved to demark end of common charger properties */
+#define CHARGER_COMMON_COUNT    CHARGER_STATUS + 1
+/**
+ * Reserved to demark downstream custom properties - use this value as the actual value may
+ * change over future versions of this API
+ */
+#define CHARGER_CUSTOM_BEGIN    CHARGER_COMMON_COUNT + 1
+
+/** Reserved to demark end of valid enum properties */
+#define CHARGER_PROP_MAX        UINT16_MAX
+
+enum charger_health {
+	CHARGER_HEALTH_UNKNOWN = 0,
+	CHARGER_HEALTH_COLD,
+	CHARGER_HEALTH_OVERHEAT,
+	CHARGER_HEALTH_GOOD,
+};
+
+enum charger_online {
+	CHARGER_ONLINE_OFFLINE = 0,
+	CHARGER_ONLINE_FIXED,
+	CHARGER_ONLINE_PROGRAMMABLE,
+};
+
+enum charger_status {
+	CHARGER_STATUS_UNKNOWN = 0,
+	CHARGER_STATUS_CHARGING,
+	CHARGER_STATUS_DISCHARGING,
+	CHARGER_STATUS_NOT_CHARGING,
+	CHARGER_STATUS_FULL,
+};
+
+struct charger_get_property {
+	/** Battery charger property to get */
+	uint16_t property_type;
+
+	/** Negative error status set by callee e.g. -ENOTSUP for an unsupported property */
+	int err;
+
+	/** Property field for getting */
+	union {
+		/* Fields have the format: */
+		/* CHARGER_PROPERTY_FIELD */
+		/* type property_field; */
+
+		/* Dynamic Charger Info */
+	/** CHARGER_HEALTH */
+	enum charger_health health;
```

Comment:
So it looks like we're abstracting away the `RES_*` statuses here. I think we're going to want to do this when creating the overarching power subsystem, but I think we want to keep the driver-level API's a little less abstract.

As a client, I'll probably want to grab the `CHARGER_SBS_STATUS` property and then check individual bits to determine health/temperature. Especially since I'd likely want to check multiple status bits at a time, I wouldn't want to have to make additional bus reads for each bit. If this ends up with us implementing only fetching a single property in this initial PR, I'm perfectly fine with that.

As an aside, in the long term I see us wanting a driver API have only a small/thin abstraction layer for making it easier to talk with the IC. This driver API then gets called by the power layer which is responsible for a lot more of the caching/abstraction of the driver to the rest of the system.

Reply:
I think I'll drop the health property for now. I brought it over from Linux sbs_charger.c, but I agree with your concerns.

The 'health' of charging would be better represented at the charging system level considering the fuel gauge and parallel charging ICs.